### PR TITLE
extractor: use daml-lf transaction Value ADT instead of bespoke ADT

### PR DESCRIPTION
**Pull request by S11001001**
extractor uses its own `LedgerValue` ADT as a pivot from the ledger-api Value type; instead, we should recycle the `Value` ADT from `daml-lf/transaction` library. This rewrites all the codecs and things to use that ADT instead, with some behavioral changes informed by the new ADT, such as proper handling of missing record field labels.

* [x] compile extractor
  * [x] alias LedgerValue
  * [x] ledger-api decoders
  * [x] JSON encoders
  * [x] doobie cases
* [x] compile extractor tests
* [x] extractor tests

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: extractor/src/main/scala/com/digitalasset/extractor/json/JsonConverters.scala [hunk 1]
```diff
@@ -89,10 +96,22 @@ object JsonConverters {
...
+  implicit def taggedDecoder[A: Decoder, T]: Decoder[A @@ T] =
+    scalaz.Tag.subst(Decoder[A])
+
+  implicit val nameEncoder: Encoder[Ref.Name] =
+    Encoder[String].contramap(identity)
+  implicit val partyEncoder: Encoder[Ref.Party] =
+    Encoder[String].contramap(identity)
+
+  // TODO SC this matches the prior behavior of JSON-ing the dates and timestamps
+  // as days-since and micros-since epoch, but maybe we'd like something else?
```
Comment (S11001001) [inline extractor/src/main/scala/com/digitalasset/extractor/json/JsonConverters.scala]: cc @bitonic @remyhaemmerle-da @leo-da for JSON-pinions

Comment (bitonic) [reply extractor/src/main/scala/com/digitalasset/extractor/json/JsonConverters.scala]: Yes, i think we should change this to use string representations, as we did with the old JSON-through-LfTypeEncoding. Feel free to change this in the next PR though.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala [hunk 2]
```diff
@@ -367,8 +367,8 @@ object ImmArray {
   implicit val immArrayInstance: Traverse[ImmArray] = new Traverse[ImmArray] {
     override def traverseImpl[F[_], A, B](immArr: ImmArray[A])(f: A => F[B])(
         implicit F: Applicative[F]): F[ImmArray[B]] = {
-      F.map(immArr.foldLeft[F[BackStack[B]]](F.point(BackStack.empty)) {
-        case (ys, x) => F.apply2(ys, f(x))(_ :+ _)
+      F.map(immArr.foldLeft[F[BackStack[B]]](F.point(BackStack.empty)) { (ys, x) =>
```
Comment (leo-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala]: not a big fan of one-liners :) really hard to follow. I would have just assigned the value returned by `foldLeft` to a var and passed it to the `map`.

File: extractor/src/main/scala/com/digitalasset/extractor/json/JsonConverters.scala [hunk 3]
```diff
@@ -3,58 +3,68 @@
...
-  implicit val recordEncoder: Encoder[Record] = record => {
-    record.fields
-      .foldLeft(JsonObject.empty) {
-        case (acc, field) =>
-          acc.+:(field.label -> field.value.asJson)
-      }
-      .asJson
-  }
+  implicit val recordEncoder: Encoder[OfCid[V.ValueRecord]] = record =>
+    if (record.fields.toSeq.forall(_._1.isDefined))
```
Comment (leo-da) [inline extractor/src/main/scala/com/digitalasset/extractor/json/JsonConverters.scala]: we convert `record.fields` to sequence twice for no reason, no matter which if/else branch we go... could reuse the value returned by the 1st `toSeq` call.

I hope we have a test case for a record with one label present `Some(x)`, and one absent `None`. Not sure if `record.fields.toSeq.map(_.asJson).asJson` will do the right thing.

File: extractor/src/main/scala/com/digitalasset/extractor/ledger/types/LedgerValue.scala [hunk 4]
```diff
@@ -61,55 +35,74 @@ object LedgerValue {
...
   }
 
   private def convertRecord(apiRecord: api.value.Record) = {
     for {
-      fields <- apiRecord.fields.toList.traverseU(_.convert)
-    } yield Record(fields)
+      tycon <- apiRecord.recordId traverseU convertIdentifier map (_.flatten)
+      // TODO SC: local RecordField may be elided, in which case convert
+      // should just produce the tuple made here
+      fields <- ImmArray(apiRecord.fields).traverseU(_.convert flatMap {
```
Comment (leo-da) [inline extractor/src/main/scala/com/digitalasset/extractor/ledger/types/LedgerValue.scala]: yeah that stuff I can't review in the browser :)

File: extractor/src/main/scala/com/digitalasset/extractor/json/JsonConverters.scala (hunk 3 above)
Comment (S11001001) [reply extractor/src/main/scala/com/digitalasset/extractor/json/JsonConverters.scala]: > we convert `record.fields` to sequence twice for no reason, no matter which if/else branch we go... could reuse the value returned by the 1st `toSeq` call.

`toSeq` is ridiculously cheap, little value in combining.

> I hope we have a test case for a record with one label present `Some(x)`, and one absent `None`.

Nope! I'll add one.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala (hunk 2 above)
Comment (S11001001) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala]: Alas Scala inference disagrees with you :) I'll rewrite it to use the syntax extension, though.
